Add azure cloud provisioner for e2e test

kartikjoshi21 opened the pull request:
Azure Provisioner to bring up aks.

yoheiueda commented:
#696 has slightly updated the provisioner related code. Could you rebase your patch?

https://github.com/confidential-containers/cloud-api-adaptor/pull/696/files#diff-627592112a06de6bfd545b8585d174910ddfc1b6b3fb03b55b480805121d258f

kartikjoshi21 commented:
> #696 has slightly updated the provisioner related code. Could you rebase your patch?
> 
> https://github.com/confidential-containers/cloud-api-adaptor/pull/696/files#diff-627592112a06de6bfd545b8585d174910ddfc1b6b3fb03b55b480805121d258f

Sure, will handle it as part of next patch. Thanks

kartikjoshi21 commented:
Tested as below
```
make BUILTIN_CLOUD_PROVIDERS="azure" build

export CLOUD_PROVIDER="azure"
export TEST_E2E_PROVISION=yes
export TEST_E2E_PROVISION_FILE=test/provisioner/provision_azure.properties

root@kartik-ThinkPad-X1-Titanium-Gen-1:/home/kartik/gitwork/gitworknew/cloud-api-adaptor# ./cluster-provisioner -action=provision
DEBU[0000] &{ResourceGroup:<nil> CloudProvider:azure SubscriptionID:d1aa957b-94f5-49ef-b29a-0178c58a7132 ClientID:e123c35e-761e-4ab9-9861-1e32ce778c2e ClientSecret:pTK8Q~SwMlvR5zplNljpHeiDMFuhe52CHQC91bQa TenantID:72f988bf-86f1-41af-91ab-2d7cd011db47 ResourceGroupName:e2e-test-rg-final ClusterName:e2e-test-cluster-final Location:eastus SshKeyID:~/.ssh/id_rsa SubnetName:e2e-test-cluster-final_subnet VnetName:e2e-test-cluster-final_vnet SubnetID: ImageID: InstanceSize:Standard_D8as_v5 NodeName:caaaks OsType:Ubuntu ResourceGroupClient:<nil> ManagedclusterClient:<nil> ManagedvnetClient:<nil> ManagedsubnetClient:<nil> ManagedaksClient:<nil> ManagedvmClient:<nil>} 
INFO[0000] Creating Resource group e2e-test-rg-final.   
INFO[0018] Successfully Created Resource group e2e-test-rg-final. 
INFO[0018] waiting for the Resource group to be available before creating vnet... 
INFO[0078] Creating  vnet e2e-test-cluster-final_vnet in resource group e2e-test-rg-final with addressPrefix 10.2.0.0/16 subnetAddressPrefix 10.2.0.0/24. 
INFO[0085] Successfully Created vnet e2e-test-cluster-final_vnet in resource group e2e-test-rg-final. 
INFO[0085] Creating cluster e2e-test-cluster-final.     
INFO[0096] Waiting for cluster e2e-test-cluster-final to be available. 
INFO[0344] Cluster e2e-test-cluster-final is available. 
INFO[0344] Sync cluster kubeconfig with current config context 
```

kartikjoshi21 commented:
Tested as follows

```
make BUILTIN_CLOUD_PROVIDERS="azure" build

export CLOUD_PROVIDER="azure"
export TEST_E2E_PROVISION_FILE="test/provisioner/provision_azure.properties"
export TEST_E2E_PROVISION="yes"

INFO[0000] Creating Resource group e2e-test-rg-final.   
INFO[0018] Successfully Created Resource group e2e-test-rg-final. 
INFO[0018] waiting for the Resource group to be available before creating vnet... 
INFO[0078] Creating  vnet e2e-test-cluster-final_vnet in resource group e2e-test-rg-final with addressPrefix 10.2.0.0/16 subnetAddressPrefix 10.2.0.0/24. 
INFO[0085] Successfully Created vnet e2e-test-cluster-final_vnet in resource group e2e-test-rg-final. 
INFO[0085] Creating cluster e2e-test-cluster-final.     
INFO[0096] Waiting for cluster e2e-test-cluster-final to be available. 
INFO[0344] Cluster e2e-test-cluster-final is available. 
INFO[0344] Sync cluster kubeconfig with current config context 
```

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
 }
 
-// AzureCloudProvisioner implements the CloudProvisioner interface for Azure.
+func CreateResourceGroup() error {
```
Any reason declare this function as public, will it be used out side of this package?

kartikjoshi21 replied:
Not intended to be public, changed the name.

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
+	return nil
+}
+
+func DeleteResourceGroup() error {
```
Any reason declare this function as public, will it be used out side of this package?

kartikjoshi21 replied:
Not intended to be public, changed the name.

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
+	return nil
+}
+
+func CreateVnetSubnet() error {
```
Any reason declare this function as public, will it be used out side of this package?

kartikjoshi21 replied:
Not intended to be public, changed the name.

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
+	return nil
+}
+
+func DeleteVNet() error {
```
Any reason declare this function as public, will it be used out side of this package?

kartikjoshi21 replied:
Not intended to be public, changed the name.

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
 }
 
-func (l *AzureCloudProvisioner) DeleteVPC(ctx context.Context, cfg *envconf.Config) error {
+func (p *AzureCloudProvisioner) DoKustomize(ctx context.Context, cfg *envconf.Config) error {
```
I think we're not using DoKustomize to customize the overlay as design in https://github.com/confidential-containers/cloud-api-adaptor/pull/668, but @wainersm may have a better understanding.

kartikjoshi21 replied:
Yes looks like with new design its not required, commented this out will remove it as soon as new mechanism is implemented.

wainersm replied:
Correct, #668 introduces a more robust mechanism. I'm trying to get it merged asap, so we can rebase this PR upon it.

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
+
+	log.Infof("Creating Resource group %s.\n", AzureProps.ResourceGroupName)
+	resourceGroup, err := AzureProps.ResourceGroupClient.CreateOrUpdate(context.Background(), AzureProps.ResourceGroupName, newRG)
+	if err != nil {
```
I guess AzureProps.ResourceGroupClient.CreateOrUpdate will not fail if rg with same name already exists?

kartikjoshi21 replied:
Yes it wont fail even if it already exist

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
+
+func DeleteResourceGroup() error {
+	log.Infof("Deleting Resource group %s.\n", AzureProps.ResourceGroupName)
+	_, err := AzureProps.ResourceGroupClient.Delete(context.Background(), AzureProps.ResourceGroupName)
```
Will AzureProps.ResourceGroupClient.Delete return `404` if delete a rg does not exist? Nice to check existing in future PR if it does.

kartikjoshi21 replied:
Yes if it tries to delete non-existing rg it returns `404`. Thanks for pointing out, resolved.

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
+
+	// Create the virtual network
+	log.Infof("Creating  vnet %s in resource group %s with addressPrefix %s subnetAddressPrefix %s.\n", AzureProps.VnetName, AzureProps.ResourceGroupName, addressPrefix, subnetAddressPrefix)
+	_, err := AzureProps.ManagedvnetClient.CreateOrUpdate(context.Background(), AzureProps.ResourceGroupName, AzureProps.VnetName, vnetParams)
```
I guess AzureProps.ManagedvnetClient.CreateOrUpdate will not fail if vnet with same name already exists?

kartikjoshi21 replied:
Yes it wont fail even if it already exist

huoqifeng commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,351 @@ package provisioner
+func DeleteVNet() error {
+	log.Infof("Deleting vnet %s in resource group %s.\n", AzureProps.VnetName, AzureProps.ResourceGroupName)
+	_, err := AzureProps.ManagedvnetClient.Delete(context.Background(), AzureProps.ResourceGroupName, AzureProps.VnetName)
+	if err != nil {
```
Will AzureProps.ManagedvnetClient.Delete return 404 if delete a vnet does not exist? Nice to check existing in future PR if it does.

kartikjoshi21 replied:
Yes if it tries to delete non-existing vnet it returns `404`. Resolved.

huoqifeng commented:
Thanks @kartikjoshi21 for adding this, I added several comments but I think several are enhancements which need not be addressed in this PR.

kartikjoshi21 commented:
> Thanks @kartikjoshi21 for adding this, I added several comments but I think several are enhancements which need not be addressed in this PR.

@huoqifeng Addressed all the comments in this PR Itself. Thanks a lot for review.

wainersm commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,326 @@ package provisioner
+		fmt.Fprintln(os.Stderr, "Unabled to find a kubeconfig file")
+		return fmt.Errorf("Unabled to find a kubeconfig file")
+	}
+	// cfg = envconf.NewWithKubeConfig(kubeconfigPath)
```
Hi @kartikjoshi21 ! I am curious why this line is comment, it didn't work to update the cfg object? That's why in `syncKubeconfig()` you had the `KUBECONFIG` environment variable updated?

kartikjoshi21 replied:
Yes it didnt work to update the cfg.
This is the reason:
```
test/provisioner/provision_azure.go:287:2: ineffectual assignment to cfg (ineffassign)
	cfg = envconf.NewWithKubeConfig(kubeconfigPath)
	^
test/provisioner/provision_azure.go:215:68: SA4009: argument cfg is overwritten before first use (staticcheck)
func (p *AzureCloudProvisioner) CreateCluster(ctx context.Context, cfg *envconf.Config) error {
                                                                   ^
```

kartikjoshi21 replied:
I will update the code later on to handle this one, Planning to submit another PR for `e2e_azure_test` related things

wainersm replied:
Got it. Thanks for the explanation @kartikjoshi21 !

wainersm reviewed (approved):
Thanks @kartikjoshi21 . Let's get this merged first then I adjust #668 . What do you think?

kartikjoshi21 commented:
> Thanks @kartikjoshi21 . Let's get this merged first then I adjust #668 . What do you think?

Sure, that will be good too, if this looks good to others too we can merge this one. @huoqifeng can you please take a look again and approve if looks good, Thanks

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+}
+
+func deleteResourceImpl() error {
+	err := deleteVNet()
```
Shouldn't just deleting the RG be enough?

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+
+func isNotFoundError(err error) bool {
+	if typedError, ok := err.(autorest.DetailedError); ok {
+		if typedError.StatusCode == 404 {
```
I think we can use the http status code `StatusNotFound` from here: https://pkg.go.dev/net/http

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+			return nil
+		}
+
+		fmt.Println("Failed to delete resource group:", err)
```
A log statement, instead of `fmt.Println`?

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+	log.Infof("Creating Resource group %s.\n", AzureProps.ResourceGroupName)
+	resourceGroup, err := AzureProps.ResourceGroupClient.CreateOrUpdate(context.Background(), AzureProps.ResourceGroupName, newRG)
+	if err != nil {
+		fmt.Println("Failed to create resource group:", err)
```
A log statement, instead of `fmt.Println`?

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+	resourceGroup, err := AzureProps.ResourceGroupClient.CreateOrUpdate(context.Background(), AzureProps.ResourceGroupName, newRG)
+	if err != nil {
+		fmt.Println("Failed to create resource group:", err)
+		return err
```
Add context to the error? I would do something like `fmt.Errorf("creating resource group: %w", err)`

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+	log.Infof("Creating  vnet %s in resource group %s with addressPrefix %s subnetAddressPrefix %s.\n", AzureProps.VnetName, AzureProps.ResourceGroupName, addressPrefix, subnetAddressPrefix)
+	_, err := AzureProps.ManagedvnetClient.CreateOrUpdate(context.Background(), AzureProps.ResourceGroupName, AzureProps.VnetName, vnetParams)
+	if err != nil {
+		return err
```
Add more context to the error

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+
+	subnet, err := AzureProps.ManagedsubnetClient.Get(context.Background(), AzureProps.ResourceGroupName, AzureProps.VnetName, AzureProps.SubnetName, "")
+	if err != nil {
+		return err
```
Add more context to the error?

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+		if isNotFoundError(err) {
+			return nil
+		}
+		fmt.Println("Failed to delete the virtual network: ", err)
+		return err
```
Similar to above please annotate the error and use log

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+func createResourceImpl() error {
+	err := createResourceGroup()
+	if err != nil {
+		return err
```
Annotate the error with information.

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+	}
+
+	log.Info("waiting for the Resource group to be available before creating vnet...")
+	time.Sleep(60 * time.Second)
```
can we have a constant instead of hardcoded value? And a comment explaining why 60 seconds were chosen?

kartikjoshi21 replied:
azure group creation take few seconds, but it worst cases it can go to 30 min, but we are creating simple resource group without any dependencies so it should get wrapped up in 1 minute and if not later step will fail saying it doesnt exist which is fine. 
Other thing we can do is use the context and add a piece of code to wait until rg with that ctx is created

mkulke replied:
I think Suraj meant to add this to the code.

surajssd commented:
Tested this one functionality wise:

```bash
✔ ~/cloud-api-adaptor [kartikjoshi21/azure-e2e-provisioner L|✚ 1…1]
$ ./cluster-provisioner -action=provision
INFO[0000] Creating Resource group e2e-test-rg.
INFO[0001] Successfully Created Resource group e2e-test-rg.
INFO[0001] waiting for the Resource group to be available before creating vnet...
INFO[0061] Creating  vnet e2e-test-cluster_vnet in resource group e2e-test-rg with addressPrefix 10.2.0.0/16 subnetAddressPrefix 10.2.0.0/24.
INFO[0063] Successfully Created vnet e2e-test-cluster_vnet in resource group e2e-test-rg.
INFO[0063] Creating cluster e2e-test-cluster.
INFO[0077] Waiting for cluster e2e-test-cluster to be available.
INFO[0322] Cluster e2e-test-cluster is available.
INFO[0322] Sync cluster kubeconfig with current config context

✔ ~/cloud-api-adaptor [kartikjoshi21/azure-e2e-provisioner L|✚ 1…1]
$ make test-e2e
go test -v -tags=azure -timeout 20m -count=1 ./test/e2e
open test/provisioner/provision_azure.properties: no such file or directory
FAIL    github.com/confidential-containers/cloud-api-adaptor/test/e2e   0.012s
FAIL
make: *** [Makefile:88: test-e2e] Error 1


✔ ~/cloud-api-adaptor [kartikjoshi21/azure-e2e-provisioner L|✚ 1…1]
$ ll test/provisioner/provision_azure.properties
.rw-rw-r-- 525 surajssd 14 Mar 23:40 test/provisioner/provision_azure.properties
```

The cluster came up fine, but I could not run the e2e test suite.


Another problem, something is wrong with the cluster I am not able to use it:

```console
➜  ~ kubectl get pods -A
Unable to connect to the server: dial tcp: lookup caa-suraj-caa-cvm0-d1aa95-vdw1je1q.hcp.eastus.azmk8s.io on 100.100.100.100:53: no such host
```

I tried it on different machines just to be sure the DNS entry hasn't been made, not sure why.

The deprovisioning worked as well:

```console
$ ./cluster-provisioner -action=deprovision
INFO[0000] Deleting Cluster e2e-test-cluster.
INFO[0001] Waiting for cluster e2e-test-cluster to be removed...
INFO[0306] Cluster e2e-test-cluster is removed.
INFO[0306] Deleting vnet e2e-test-cluster_vnet in resource group e2e-test-rg.
INFO[0307] Successfully Deleted vnet e2e-test-cluster_vnet in resource group e2e-test-rg.
INFO[0307] Deleting Resource group e2e-test-rg.
INFO[0308] Successfully Deleted Resource group e2e-test-rg.
```

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+		return err
+	}
+
+	// check if cluster is ready after creation
+	clusterReady := false
+	waitMinutes := 10
+	log.Infof("Waiting for cluster %s to be available.\n", AzureProps.ClusterName)
+	for i := 0; i <= waitMinutes; i++ {
+		ready, err := isAzureClusterReady(AzureProps.ResourceGroupName, AzureProps.ClusterName)
+		if err != nil {
+			log.Warnf("Err %s happened when retrieve cluster, try again...\n", err)
+			continue
+		}
+		if ready {
+			log.Infof("Cluster %s is available.\n", AzureProps.ClusterName)
+			clusterReady = true
+			break
+		}
+
+		time.Sleep(60 * time.Second)
+	}
```
I think I would replace this custom polling mechanism would something like the following provided by the package: "k8s.io/apimachinery/pkg/util/wait"

```go
if err = wait.PollImmediate(retryInterval, timeout, func() (done bool, err error) {
	// gotta try again
	return false, nil

	// error occurred while waiting
	return false, err

	// Successful outcome
	return true, nil
}); err != nil {
}
```

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+		return err
+	}
+
+	clusterRemoved := false
+	waitMinutes := 10
+	log.Infof("Waiting for cluster %s to be removed...\n", AzureProps.ClusterName)
+	for i := 0; i <= waitMinutes; i++ {
+		found, err := clusterExists(AzureProps.ResourceGroupName, AzureProps.ClusterName)
+		if err != nil {
+			log.Warnf("Err %s happened when retrieve cluster, try again...\n", err)
+			continue
+		}
+		if !found {
+			log.Infof("Cluster %s is removed.\n", AzureProps.ClusterName)
+			clusterRemoved = true
+			break
+		}
+		time.Sleep(60 * time.Second)
+	}
+
+	if !clusterRemoved {
+		return fmt.Errorf("Cluster %s was not removed completely in %d minutes.\n", AzureProps.ClusterName, waitMinutes)
+	}
```
Again the custom waiting mechanism can be replaced.

surajssd commented on test/provisioner/provision_azure_initializer.go:
```diff
@@ -0,0 +1,165 @@
+
+var AzureProps = &AzureProperties{}
+
+func initazureProperties(properties map[string]string) error {
```
```suggestion
func initAzureProperties(properties map[string]string) error {
```

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+			VMSize:             &AzureProps.InstanceSize,
+			Mode:               containerservice.AgentPoolModeSystem,
+			OsType:             containerservice.OSType(AzureProps.OsType),
+			EnableNodePublicIP: to.BoolPtr(true),
```
Why do we need public ip for the node?

kartikjoshi21 replied:
We dont need it actually.

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+	agentPoolProperties := []containerservice.ManagedClusterAgentPoolProfile{
+		{
+			Name:               &AzureProps.NodeName,
+			Count:              to.Int32Ptr(2),
```
Do we need two workers?

surajssd commented on test/provisioner/provision_azure_initializer.go:
```diff
@@ -0,0 +1,165 @@
+
+	AzureProps.VnetName = AzureProps.ClusterName + "_vnet"
+	AzureProps.SubnetName = AzureProps.ClusterName + "_subnet"
+	AzureProps.InstanceSize = "Standard_D8as_v5"
```
I think this is a very big machine for running a CAA DS pod.

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
 }
 
-func (l *AzureCloudProvisioner) UploadPodvm(imagePath string, ctx context.Context, cfg *envconf.Config) error {
+//func (p *AzureCloudProvisioner) DoKustomize(ctx context.Context, cfg *envconf.Config) error {
+//	overlayFile := "../../install/overlays/azure/kustomization.yaml"
+//	overlayFileBak := "../../install/overlays/azure/kustomization.yaml.bak"
+//	//runtimeCri := "imagepullsecret.yaml"
+//	err := os.Rename(overlayFile, overlayFileBak)
+//	if err != nil {
+//		return err
+//	}
+//
+//	input, err := os.ReadFile(overlayFileBak)
+//	if err != nil {
+//		return err
+//	}
+//
+//	replacer := strings.NewReplacer("CLOUD_PROVIDER=\"\"", "CLOUD_PROVIDER=\""+AzureProps.CloudProvider+"\"",
+//		"AZURE_SUBSCRIPTION_ID=\"\"", "AZURE_SUBSCRIPTION_ID=\""+AzureProps.SubscriptionID+"\"",
+//		"AZURE_REGION=\"\"", "AZURE_REGION=\""+AzureProps.Location+"\"",
+//		"AZURE_INSTANCE_SIZE=\"\"", "AZURE_INSTANCE_SIZE=\""+AzureProps.InstanceSize+"\"",
+//		"AZURE_SUBNET_ID=\"\"", "AZURE_SUBNET_ID=\""+AzureProps.SubnetID+"\"",
+//		"AZURE_RESOURCE_GROUP=\"\"", "AZURE_RESOURCE_GROUP=\""+AzureProps.ResourceGroupName+"\"",
+//		"AZURE_CLIENT_ID=\"\"", "AZURE_CLIENT_ID=\""+AzureProps.ClientID+"\"",
+//		"AZURE_CLIENT_SECRET=\"\"", "AZURE_CLIENT_SECRET=\""+AzureProps.ClientSecret+"\"",
+//		"AZURE_TENANT_ID=\"\"", "AZURE_TENANT_ID=\""+AzureProps.TenantID+"\"",
+//		"AZURE_IMAGE_ID=\"\"", "AZURE_IMAGE_ID=\""+AzureProps.ImageID+"\"")
+//
+//	output := replacer.Replace(string(input))
+//
+//	if err = os.WriteFile(overlayFile, []byte(output), 0666); err != nil {
+//		return err
+//	}
+//
+//	return nil
+//}
```
Should be removed?

surajssd commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+
+func (p *AzureCloudProvisioner) UploadPodvm(imagePath string, ctx context.Context, cfg *envconf.Config) error {
+	log.Trace("UploadPodvm()")
+	log.Trace("Image is uploaded via packer in case of azure")
```
Where do we trigger packer in this case?

kartikjoshi21 replied:
I will handle that in next PR where will get whole e2e test working along with Wainer's PR on install overlays.

surajssd commented on test/provisioner/provision_azure_initializer.go:
```diff
@@ -0,0 +1,165 @@
+}
+
+func initAzureLogger() {
+	level := os.Getenv("LOG_LEVEL")
```
I wonder if the library automatically picks up the value and we don't have to do this switch case?

surajssd commented on test/provisioner/provision_azure_initializer.go:
```diff
@@ -0,0 +1,165 @@
+	AzureProps.NodeName = "caaaks"
+	AzureProps.OsType = "Ubuntu"
+
+	log.Debugf("%+v", AzureProps)
```
This can leak secrets when debug is enabled. Is there a way to disable that?

kartikjoshi21 replied:
I think its better to remove this debug

surajssd commented on test/provisioner/provision_azure_initializer.go:
```diff
@@ -0,0 +1,165 @@
+
+	log.Debugf("%+v", AzureProps)
+
+	if len(AzureProps.SubscriptionID) <= 0 {
```
For a string even this is an easy way to check if it empty!

```suggestion
	if AzureProps.SubscriptionID == "" {
```

surajssd commented on test/provisioner/provision_azure_initializer.go:
```diff
@@ -0,0 +1,165 @@
+	log.Trace("initManagedClients()")
+	authorizer, err := auth.NewAuthorizerFromEnvironment()
+	if err != nil {
+		return err
```
Annotate error with `fmt.Errorf` ?

surajssd replied:
@kartikjoshi21 I think this was missed?

mkulke commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+
+	ManagedClusterProperties := &containerservice.ManagedClusterProperties{
+		ServicePrincipalProfile: servicePrincipalProfile,
+		DNSPrefix:               to.StringPtr("caa-dns-prefix"),
```
this should probably just be "caa"?

mkulke commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+		if isNotFoundError(err) {
+			return false, nil
+		}
+		return false, fmt.Errorf("failed to get cluster: %v", err)
```
should be an error log?

mkulke commented on test/provisioner/provision_azure.go:
```diff
@@ -8,37 +8,355 @@ package provisioner
+func isAzureClusterReady(resourceGroupName string, clusterName string) (bool, error) {
+	cluster, err := AzureProps.ManagedaksClient.Get(context.Background(), resourceGroupName, clusterName)
+	if err != nil {
+		return false, fmt.Errorf("failed to get cluster: %v", err)
```
should be a logger call?

mkulke commented on test/provisioner/provision_azure.properties:
```diff
@@ -0,0 +1,11 @@
+AZURE_SUBSCRIPTION_ID=${AZURE_SUBSCRIPTION_ID}
+# use az ad sp create-for-rbac   --role Contributor   --scopes "/subscriptions/$SUBSCRIPTION_ID" --query "{ client_id: appId, client_secret: password, tenant_id: tenant }"
+AZURE_CLIENT_ID=${AZURE_CLIENT_ID}
+AZURE_CLIENT_SECRET=${AZURE_CLIENT_SECRET}
```
not sure how those variables are sourced, but you might want to quote the secret, sometimes there are characters like whitespace in them.

kartikjoshi21 commented:
> Tested this one functionality wise:
> 
> ```shell
> ✔ ~/cloud-api-adaptor [kartikjoshi21/azure-e2e-provisioner L|✚ 1…1]
> $ ./cluster-provisioner -action=provision
> INFO[0000] Creating Resource group e2e-test-rg.
> INFO[0001] Successfully Created Resource group e2e-test-rg.
> INFO[0001] waiting for the Resource group to be available before creating vnet...
> INFO[0061] Creating  vnet e2e-test-cluster_vnet in resource group e2e-test-rg with addressPrefix 10.2.0.0/16 subnetAddressPrefix 10.2.0.0/24.
> INFO[0063] Successfully Created vnet e2e-test-cluster_vnet in resource group e2e-test-rg.
> INFO[0063] Creating cluster e2e-test-cluster.
> INFO[0077] Waiting for cluster e2e-test-cluster to be available.
> INFO[0322] Cluster e2e-test-cluster is available.
> INFO[0322] Sync cluster kubeconfig with current config context
> 
> ✔ ~/cloud-api-adaptor [kartikjoshi21/azure-e2e-provisioner L|✚ 1…1]
> $ make test-e2e
> go test -v -tags=azure -timeout 20m -count=1 ./test/e2e
> open test/provisioner/provision_azure.properties: no such file or directory
> FAIL    github.com/confidential-containers/cloud-api-adaptor/test/e2e   0.012s
> FAIL
> make: *** [Makefile:88: test-e2e] Error 1
> 
> 
> ✔ ~/cloud-api-adaptor [kartikjoshi21/azure-e2e-provisioner L|✚ 1…1]
> $ ll test/provisioner/provision_azure.properties
> .rw-rw-r-- 525 surajssd 14 Mar 23:40 test/provisioner/provision_azure.properties
> ```
> 
> The cluster came up fine, but I could not run the e2e test suite.
> 
> Another problem, something is wrong with the cluster I am not able to use it:
> 
> ```
> ➜  ~ kubectl get pods -A
> Unable to connect to the server: dial tcp: lookup caa-suraj-caa-cvm0-d1aa95-vdw1je1q.hcp.eastus.azmk8s.io on 100.100.100.100:53: no such host
> ```
> 
> I tried it on different machines just to be sure the DNS entry hasn't been made, not sure why.
> 
> The deprovisioning worked as well:
> 
> ```
> $ ./cluster-provisioner -action=deprovision
> INFO[0000] Deleting Cluster e2e-test-cluster.
> INFO[0001] Waiting for cluster e2e-test-cluster to be removed...
> INFO[0306] Cluster e2e-test-cluster is removed.
> INFO[0306] Deleting vnet e2e-test-cluster_vnet in resource group e2e-test-rg.
> INFO[0307] Successfully Deleted vnet e2e-test-cluster_vnet in resource group e2e-test-rg.
> INFO[0307] Deleting Resource group e2e-test-rg.
> INFO[0308] Successfully Deleted Resource group e2e-test-rg.
> ```

@surajssd yes this PR is supposed to get provisioner working, other than that I will submit a new PR for e2e test to work with this. It requires install overlay side things as well.

wainersm commented:
@kartikjoshi21 https://github.com/confidential-containers/cloud-api-adaptor/pull/668 just got approved but I didn't merge it yet, if I do so then this PR will need changes. And if we merge this one first, the other way around :)

Mind if we get 668 in first as it is touches the core of the framework and others providers implementations depends on it? I will assist you on any needed changes on this PR.

wainersm commented:
> > @kartikjoshi21 #668 just got approved but I didn't merge it yet, if I do so then this PR will need changes. And if we merge this one first, the other way around :)
> > Mind if we get 668 in first as it is touches the core of the framework and others providers implementations depends on it? I will assist you on any needed changes on this PR.
> 
> @wainersm If 668 is ready lets merge it i will make changes to this PR accordingly. Thanks !!

668 is merged now.

huoqifeng reviewed (approved):
/lgtm, thank you! @kartikjoshi21

surajssd commented:
The deprovisioning failed for me, twice:

```console
$ ./cluster-provisioner -action=deprovision
INFO[0000] Deleting Cluster e2e-test-cluster.
INFO[0001] Waiting for cluster e2e-test-cluster to be removed...
ERRO[0302] failed to get cluster: containerservice.ManagedClustersClient#Get: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The Resource 'Microsoft.ContainerService/managedClusters/e2e-test-cluster' under resource group 'e2e-test-rg' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix"
waiting for cluster to be deleted failed to wait for condition: failed to get cluster: containerservice.ManagedClustersClient#Get: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The Resource 'Microsoft.ContainerService/managedClusters/e2e-test-cluster' under resource group 'e2e-test-rg' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix"
```

I think we need to add some form of cleanup, this allows leftover resources to stay back. Aggressively try to delete the resource group a couple of times.

--- 

The kubeconfig isn't configured properly. To test this, I deleted the kubeconfig file provisioned the cluster and still no kubeconfig file was created.

```bash
$ rm -rf ~/.kube/config

$ ./cluster-provisioner -action=provision
INFO[0000] Creating Resource group e2e-test-rg1.
INFO[0006] Successfully Created Resource group e2e-test-rg1.
INFO[0006] waiting for the Resource group to be available before creating vnet...
INFO[0066] Creating  vnet e2e-test-cluster_vnet in resource group e2e-test-rg1 with addressPrefix 10.2.0.0/16 subnetAddressPrefix 10.2.0.0/24.
INFO[0069] Successfully Created vnet e2e-test-cluster_vnet in resource group e2e-test-rg1.
INFO[0069] Creating cluster e2e-test-cluster.
INFO[0081] Waiting for cluster e2e-test-cluster to be available.
INFO[0323] Sync cluster kubeconfig with current config context

$ ls ~/.kube/config
"/home/surajaz/.kube/config": No such file or directory (os error 2)
```

kartikjoshi21 commented on test/provisioner/provision_azure_initializer.go:
```diff
@@ -0,0 +1,156 @@
+	log.SetLevel(level)
+}
+
+type AzureProperties struct {
```
We need this struct in e2e package as well. Other thing which can be done to have a getter kind of thing but it is one of the same thing.

kartikjoshi21 commented on test/provisioner/provision_azure.go:
```diff
@@ -26,30 +194,120 @@ type AzureInstallOverlay struct {
-	return nil
+func (p *AzureCloudProvisioner) CreateVPC(ctx context.Context, cfg *envconf.Config) error {
+	log.Trace("CreateVPC()")
+	return createResourceImpl()
```
Its just something i followed from imbcloud structure :) But its just means it implements.

surajssd replied:
I asked because this is not an interface, so was wondering if we should call that or just call it `createResoucrces`.

kartikjoshi21 commented on test/provisioner/provision_azure.go:
```diff
@@ -8,15 +8,183 @@ package provisioner
+		return fmt.Errorf("creating resource group: %v", err)
+	}
+
+	// rg creation takes few seconds to complete keeping it as 60 second to be on safe side.
+	const sleeptime = time.Duration(60) * time.Second
+	log.Info("waiting for the Resource group to be available before creating vnet...")
+	time.Sleep(sleeptime)
```
Yess i have plans to consolidate it and provide wait for resource group creation in next PR

kartikjoshi21 commented:
> The deprovisioning failed for me, twice:
> 
> ```
> $ ./cluster-provisioner -action=deprovision
> INFO[0000] Deleting Cluster e2e-test-cluster.
> INFO[0001] Waiting for cluster e2e-test-cluster to be removed...
> ERRO[0302] failed to get cluster: containerservice.ManagedClustersClient#Get: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The Resource 'Microsoft.ContainerService/managedClusters/e2e-test-cluster' under resource group 'e2e-test-rg' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix"
> waiting for cluster to be deleted failed to wait for condition: failed to get cluster: containerservice.ManagedClustersClient#Get: Failure responding to request: StatusCode=404 -- Original Error: autorest/azure: Service returned an error. Status=404 Code="ResourceNotFound" Message="The Resource 'Microsoft.ContainerService/managedClusters/e2e-test-cluster' under resource group 'e2e-test-rg' was not found. For more details please go to https://aka.ms/ARMResourceNotFoundFix"
> ```
> 
> I think we need to add some form of cleanup, this allows leftover resources to stay back. Aggressively try to delete the resource group a couple of times.
> 
> The kubeconfig isn't configured properly. To test this, I deleted the kubeconfig file provisioned the cluster and still no kubeconfig file was created.
> 
> ```shell
> $ rm -rf ~/.kube/config
> 
> $ ./cluster-provisioner -action=provision
> INFO[0000] Creating Resource group e2e-test-rg1.
> INFO[0006] Successfully Created Resource group e2e-test-rg1.
> INFO[0006] waiting for the Resource group to be available before creating vnet...
> INFO[0066] Creating  vnet e2e-test-cluster_vnet in resource group e2e-test-rg1 with addressPrefix 10.2.0.0/16 subnetAddressPrefix 10.2.0.0/24.
> INFO[0069] Successfully Created vnet e2e-test-cluster_vnet in resource group e2e-test-rg1.
> INFO[0069] Creating cluster e2e-test-cluster.
> INFO[0081] Waiting for cluster e2e-test-cluster to be available.
> INFO[0323] Sync cluster kubeconfig with current config context
> 
> $ ls ~/.kube/config
> "/home/surajaz/.kube/config": No such file or directory (os error 2)
> ```

@surajssd yes there was one status code which was not working fine. Resolved all the comments

```
root@kartik-ThinkPad-X1-Titanium-Gen-1:/home/kartik/gitwork/gitworknew/cloud-api-adaptor# ./cluster-provisioner -action=provision
INFO[0000] Creating Resource group e2e-test-rg.         
INFO[0003] Successfully Created Resource group e2e-test-rg. 
INFO[0003] waiting for the Resource group to be available before creating vnet... 
INFO[0063] Creating  vnet e2e-test-cluster_vnet in resource group e2e-test-rg with addressPrefix 10.2.0.0/16 subnetAddressPrefix 10.2.0.0/24. 
INFO[0069] Successfully Created vnet e2e-test-cluster_vnet in resource group e2e-test-rg. 
INFO[0069] Creating cluster e2e-test-cluster.           
INFO[0081] Waiting for cluster e2e-test-cluster to be available. 
DEBU[0081] isAzureClusterReady()                        
DEBU[0143] isAzureClusterReady()                        
DEBU[0203] isAzureClusterReady()                        
DEBU[0263] isAzureClusterReady()                        
DEBU[0323] isAzureClusterReady()                        
INFO[0325] Sync cluster kubeconfig with current config context 
root@kartik-ThinkPad-X1-Titanium-Gen-1:/home/kartik/gitwork/gitworknew/cloud-api-adaptor# ls -lrt $HOME/.kube/config 
-rw------- 1 root docker 9728 Mar 16 15:42 /root/.kube/config
root@kartik-ThinkPad-X1-Titanium-Gen-1:/home/kartik/gitwork/gitworknew/cloud-api-adaptor# date
Thursday 16 March 2023 03:43:45 PM IST

root@kartik-ThinkPad-X1-Titanium-Gen-1:/home/kartik/gitwork/gitworknew/cloud-api-adaptor# ./cluster-provisioner -action=deprovision
INFO[0000] Deleting Cluster e2e-test-cluster.           
INFO[0013] Waiting for cluster e2e-test-cluster to be removed... 
DEBU[0013] isClusterDeleted()                           
DEBU[0075] isClusterDeleted()                           
DEBU[0135] isClusterDeleted()                           
DEBU[0195] isClusterDeleted()                           
DEBU[0255] isClusterDeleted()                           
DEBU[0315] isClusterDeleted()                           
INFO[0315] Deleting Resource group e2e-test-rg.         
INFO[0321] Successfully Deleted Resource group e2e-test-rg. 
```

Provision and deprovision works fine. Regarding running e2e test that will be taken care of in next PR, which will consist of adding install overlay part. Thanks

kartikjoshi21 commented on test/provisioner/provision_azure.properties:
```diff
@@ -0,0 +1,11 @@
+AZURE_CLIENT_SECRET=${AZURE_CLIENT_SECRET}
+AZURE_TENANT_ID=${AZURE_TENANT_ID}
+
+RESOURCE_GROUP_NAME="e2e-test-rg"
```
We can think of either adding suffix to the rg name or leave it to users to change it, will discuss it and make the change accordingly.

surajssd reviewed (changes requested):
Functionality wise things work. Great work Kartik, last few bits and then it is good to go:


I see that the binary is not in the `.gitignore` file:

```console
$ git status
On branch pr_680_upstream_834828478
Untracked files:
  (use "git add <file>..." to include in what will be committed)
        cluster-provisioner

nothing added to commit but untracked files present (use "git add" to track)
```

kartikjoshi21 commented:
> ```
> cluster-provisioner
> ```

Thankyou @surajssd for review, made all the mentioned changes.
